### PR TITLE
Add Checkbox form component

### PR DESCRIPTION
**Pull request by rezrah**
# Checkbox  ✅  

Adds a native `Checkbox` component that provides a react abstraction over a native HTML checkbox input 

Adapted from the Primer design guidelines for checkboxes

**Before:**
`<input type="checkbox" />`

**After:** 
`<Checkbox />`
### Try it here
Docs: [https://primer-components-eo45yjh4t-primer.vercel.app/react/Checkbox](https://primer-components-eo45yjh4t-primer.vercel.app/react/Checkbox)
Storybook: https://primer-components-eo45yjh4t-primer.vercel.app/react/storybook?path=/story/forms-checkbox--default

### Screenshots

Screenshot 2021-11-19 at 15 22 01
Screenshot 2021-11-19 at 15 21 08
Screenshot 2021-11-19 at 15 21 44
Screenshot 2021-11-19 at 15 20 51

### Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation
- [x] Tested in Chrome
- [x] Tested in Firefox
- [x] Tested in Safari
- [x] Tested in Edge

Resolves https://github.com/github/primer/issues/467
Contributes towards: https://github.com/github/primer/issues/432

### CODE REVIEW

**Review by mperrotti — COMMENTED** [review]
Since this is still in progress, I kept my feedback limited to the visual design and component API.

I'll review the code once we're further along.

File: src/Checkbox.tsx [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+  /**
+   * Apply disabled state for the checkbox
+   */
+  disabled?: boolean
+
+  /**
+   * Force container to take up full width of parent
+   */
+  block?: boolean
```
[the remark this answers is not in the document]
Comment (rezrah) [reply src/Checkbox.tsx]: removed 👍 , it was on the previous iteration to help with label layout but feels safe to move this to form layout instead

File: src/Checkbox.tsx [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  checked?: boolean
+
+  /**
+   * Optional callback for when the checkbox is clicked
+   */
+  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
+  /**
+   * An adjacent label for the checkbox
+   */
+  label?: string
```
[the remark this answers is not in the document]
Comment (rezrah) [reply src/Checkbox.tsx]: Agree 👍 . Latest iteration makes this a standalone component. I do think we should make this public though, as it's better to use this Component than have other Primer users implement their own checkboxes outside of `CheckboxField`

File: src/Checkbox.tsx [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+  position: absolute;
+  clip: rect(1px, 1px, 1px, 1px);
+  padding: 0;
+  border: 0;
+  height: 1px;
+  width: 1px;
+  overflow: hidden;
+`
+
+const StyledCheckbox = styled.label<Checkbox>`
```
[the remark this answers is not in the document]
Comment (rezrah) [reply src/Checkbox.tsx]: Adapted from Figma. Admittedly colours were all wrong. Component has recently switched over to native styling now, please feel free to re-review @mperrotti.

File: src/Checkbox.tsx [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+type Container = {
+  block?: boolean
+  labelText?: string
+  sx?: SxProps
+} & HTMLAttributes<HTMLDivElement>
+
+const StyledComponentContainer = styled.div<Container>`
+  display: inline-block;
+  align-items: center;
```
[the remark this answers is not in the document]
Comment (rezrah) [reply src/Checkbox.tsx]: No longer an issue, due to removing the label.

File: src/Checkbox.tsx [hunk 5]
```diff
@@ -0,0 +1,230 @@
+import styled, {css} from 'styled-components'
+import React, {InputHTMLAttributes, HTMLAttributes, ReactElement, useEffect, useRef} from 'react'
+import {Text, TextProps} from './'
+import sx, {BetterSystemStyleObject as SxProps} from './sx'
+import {IconProps, CheckIcon, DashIcon} from '@primer/octicons-react'
+import {COMMON, get} from './constants'
+
+export type Checkbox = {
```
[the remark this answers is not in the document]
Comment (rezrah) [reply src/Checkbox.tsx]: Having opted to use a native checkbox, we're not going to be able to do this. We can however, add add some aria-labels to indicate the field is invalid. In order to get the danger border, we'll need to apply custom styling, so I'd suggest we hold off until we have all other states covered too. Perhaps when we move to `beta` for this component?

File: docs/content/Checkbox.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  <Checkbox indeterminate checked={true} label="Indeterminate" />
+  <Checkbox indeterminate disabled checked={true} label="Indeterminate + Disabled" />
+</>
+```
+
+## Component props
+
+Native `<input>` attributes are forwarded to the underlying React `input` component and are not listed below.
+
+| Name          | Type    |  Default  | Description                                                                                                                                   |
```
[the remark this answers is not in the document]
Comment (rezrah) [reply docs/content/Checkbox.md]: It's the same table as the other component docs, but happy to switch over to the Props component. I admittedly didn't realise we had one 🤔 .

File: src/Checkbox.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (rezrah) [reply src/Checkbox.tsx]: Fixed 👍 , good shout.

File: src/Checkbox.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  ref?: React.RefObject<HTMLInputElement>
+} & InputHTMLAttributes<HTMLInputElement> &
+  SxProp
+
+const StyledCheckbox = styled.input`
+  cursor: pointer;
+
+  ${props => props.disabled && `cursor: not-allowed;`}
+
+  ${COMMON};
```
Comment (mperrotti) [inline src/Checkbox.tsx:28]: I don't think we need to allow `COMMON` props

File: src/Checkbox.tsx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    }, [indeterminate, checked])
+
+    return (
+      <StyledCheckbox
+        type="checkbox"
+        disabled={disabled}
+        aria-disabled={disabled ? 'true' : 'false'}
+        ref={ref || checkboxRef}
+        checked={indeterminate ? false : checked}
+        aria-checked={indeterminate ? 'mixed' : checked ? 'true' : 'false'}
```
Comment (mperrotti) [inline src/Checkbox.tsx:52]: Nice

File: src/Checkbox.tsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
+import styled from 'styled-components'
+import React, {InputHTMLAttributes, ReactElement, useLayoutEffect, useRef} from 'react'
+import sx, {SxProp} from './sx'
+import {COMMON} from './constants'
+
+export type CheckboxProps = {
```
Comment (mperrotti) [inline src/Checkbox.tsx:6]: We should add a prop for validation status. Something like:
```
validationStatus?: 'error' | 'success'
```

This would style the input to look like a "success" or "error", and set `aria-invalid="true"` if it's an error.

cc @alliethu - am I correct in thinking that `aria-invalid="true"` should be set when the field fails validation?

File: src/Checkbox.tsx [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  ${props => props.disabled && `cursor: not-allowed;`}
+
+  ${COMMON};
+  ${sx}
+`
+
+/**
+ * An accessible, native checkbox component
+ */
+const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
```
Comment (mperrotti) [inline src/Checkbox.tsx:35]: Why do we need to use `forwardRef`?

File: src/index.ts [hunk 11]
```diff
@@ -169,4 +169,7 @@ export type {TruncateProps} from './Truncate'
 export {default as UnderlineNav} from './UnderlineNav'
 export type {UnderlineNavProps, UnderlineNavLinkProps} from './UnderlineNav'
 
+export {default as Checkbox} from './Checkbox'
```
Comment (mperrotti) [inline src/index.ts:172]: Any ideas how we could guard against engineers rendering a `<Checkbox />` without a corresponding `<label />`?

I have some ideas, but it would be much easier to show over Zoom instead of text and pseudocode. Let me know if you want to chat on Monday.

File: docs/content/Checkbox.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+)
+}
+
+<Note variant="warning">
+Checkbox components should always be accompanied by a corrosponding label to improve support for assistive technologies.
+</Note>
+
+## Indeterminate example
+
+An `indeterminate` checkbox state can be used if the inputs value is in-between on or off. This can be useful in situations where you are required to display an incomplete state, that is dependent on other external selections to force a binary decision of the parent checkbox.
```
Comment (mperrotti) [inline docs/content/Checkbox.md:58]: Instead of saying "in-between on or off", I'd recommend something like "is neither true nor false"

File: src/Checkbox.tsx (hunk 10 above)
Comment (rezrah) [reply src/Checkbox.tsx:43]: It's there to make the checkbox more flexible when used in uncontrolled mode.

File: src/Checkbox.tsx (hunk 9 above)
Comment (rezrah) [reply src/Checkbox.tsx:6]: nice, i've added in the aria roles now. We're not going to be able to easily style this like you suggested though, as it's a native checkbox and we've opting for system styles. Perhaps that part can come in on the next iteration.

Comment (mperrotti) [reply src/Checkbox.tsx:6]: Do we have an issue created for adding validation styles?

File: src/index.ts (hunk 11 above)
Comment (mperrotti) [reply src/index.ts:172]: Chatted with @rezrah about this: we're just going to check for an associated label with a linter instead of doing anything in the runtime.

**Review by mperrotti — APPROVED** [review]

File: src/Checkbox.tsx [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+    return (
+      <StyledCheckbox
+        type="checkbox"
+        disabled={disabled}
+        aria-disabled={disabled ? 'true' : 'false'}
+        ref={ref || checkboxRef}
+        checked={indeterminate ? false : checked}
+        aria-checked={indeterminate ? 'mixed' : checked ? 'true' : 'false'}
+        sx={{...sxProp}}
```
Comment (colebemis) [inline src/Checkbox.tsx:62]: Why do we need to spread `...sxProp` here?
```suggestion
        sx={sxProp}
```

File: src/Checkbox.tsx [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * An accessible, native checkbox component
+ */
+const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
+  (
+    {checked, indeterminate, disabled, sx: sxProp, required, validationStatus, ...rest}: CheckboxProps,
+    ref
+  ): ReactElement => {
+    const checkboxRef = useRef<HTMLInputElement>(null)
```
Comment (colebemis) [inline src/Checkbox.tsx:46]: Would it make sense for us to use `useProvedRefOrCreate()` here? https://github.com/primer/react/blob/main/src/hooks/useProvidedRefOrCreate.ts

File: docs/content/Checkbox.md [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+import { useState } from 'react'
+import { useMDXScope } from "gatsby-plugin-mdx/context";
+
+export const MyDynamicButton = () => {
+const [count, setCount] = useState(0)
+return (
+<button onClick={() => setCount(count + 1)}>Clicked {count} times!</button>
+)
+}
```
Comment (colebemis) [inline docs/content/Checkbox.md:53]: What is this for?

File: docs/content/Checkbox.md [hunk 16]
```diff
@@ -0,0 +1,127 @@
+---
+title: Checkbox
+status: Alpha
+source: https://github.com/primer/react/blob/main/src/Checklist.tsx
+---
```
Comment (colebemis) [inline docs/content/Checkbox.md:5]: We can also add a link to storybook here:
```suggestion
source: https://github.com/primer/react/blob/main/src/Checklist.tsx
storybook: ...
---
```

File: docs/content/Checkbox.md [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+</>
+```
+
+## Component props
+
+Native `<input>` attributes are forwarded to the underlying React `input` component and are not listed below.
+
+| Name           | Type    |  Default  | Description                                                                                                                                             |
+| :------------- | :------ | :-------: | :------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| checked        | Boolean | undefined | Optional. Modifies true/false value of the native checkbox                                                                                              |
```
Comment (colebemis) [inline docs/content/Checkbox.md:94]: Should we document the `onChange` prop? That seems like an essential piece of the "controlled" mode

File: docs/content/Checkbox.md (hunk 15 above)
Comment (rezrah) [reply docs/content/Checkbox.md]: Wow.. thanks @colebemis... that's a hangover from some mdx testing. Removed now 👍

File: src/Checkbox.tsx (hunk 13 above)
Comment (rezrah) [reply src/Checkbox.tsx]: Fixed.

File: src/Checkbox.tsx (hunk 14 above)
Comment (rezrah) [reply src/Checkbox.tsx]: Didn't realise we had this. Awesome. I've swapped it in and checked it works the same 👍

File: docs/content/Checkbox.md (hunk 16 above)
Comment (rezrah) [reply docs/content/Checkbox.md]: 👍 Added

File: docs/content/Checkbox.md (hunk 17 above)
Comment (rezrah) [reply docs/content/Checkbox.md]: I've added it for convenience. Can you verify it's okay please @colebemis?

File: docs/content/Checkbox.md [hunk 18]
```diff
@@ -0,0 +1,119 @@
+---
+title: Checkbox
+status: Alpha
+source: https://github.com/primer/react/blob/main/src/Checklist.tsx
+storybook: '/react/storybook?path=/story/forms-checkbox--default'
+---
+
+import {ComponentChecklist} from '../src/component-checklist'
+
+`Checkbox` is a form component that applies Primer design language to a native HTML checkbox input element.
```
Comment (colebemis) [inline docs/content/Checkbox.md:10]: Could we move this into the `description` frontmatter and reword it to be a little more actionable (see other [component descriptions](https://github.com/primer/component-metadata/blob/main/metadata.json)):

```
---
title: Checkbox
description: Use checkboxes to ...
---
```

File: docs/content/Checkbox.md [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+
+## Default example
+
+The `Checkbox` component can be used in controlled and uncontrolled modes.
+
+```jsx live
+<>
+  <Box as="form" p={3} pt={0} sx={{display: 'flex', alignItems: 'center'}}>
```
Comment (colebemis) [inline docs/content/Checkbox.md:18]: `Box` techincally supports both system props and the `sx` prop but it feels a little messy to use them both at the same time. What if we moved all the styling into the `sx` prop in these examples?

```suggestion
  <Box as="form" sx={{p: 3, pt: 0, display: 'flex', alignItems: 'center'}}>
```

File: docs/content/Checkbox.md (hunk 17 above)
Comment (colebemis) [reply docs/content/Checkbox.md]: Looks great 👍

File: docs/content/Checkbox.md (hunk 19 above)
Comment (rezrah) [reply docs/content/Checkbox.md]: Fair. That's done now 👍

File: docs/content/Checkbox.md (hunk 18 above)
Comment (rezrah) [reply docs/content/Checkbox.md]: Moved and content updated 👍

**Review by colebemis — APPROVED** [review]
🚀

File: src/Checkbox.tsx (hunk 9 above)
Comment (rezrah) [reply src/Checkbox.tsx:6]: Not yet, but i'll create one and link it to https://github.com/github/primer/issues/467

Comment (mperrotti) [reply src/Checkbox.tsx:6]: Issue for adding validation styles: https://github.com/github/primer/issues/495
